<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="TOML files" href="toml.html" /><link rel="prev" title="Welcome to the readfish documentation!" href="index.html" />

    <!-- Generated with Sphinx 7.3.7 and Furo 2024.05.06 -->
        <title>Getting Started - readfish 2024.2.0</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=387cc868" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    <link rel="stylesheet" type="text/css" href="_static/css/custom.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --font-stack: Roboto, -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;
  --font-stack--monospace: 'Ubuntu Mono', monospace;
  --code-font-size: 90%;
  --color-api-background: #f8f9fb;
  --color-highlight-on-target: transparent;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  --color-api-background: #1a1c1e;
  --color-highlight-on-target: transparent;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  --color-api-background: #1a1c1e;
  --color-highlight-on-target: transparent;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" /
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">readfish 2024.2.0</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="_static/readfish_light.png" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="_static/readfish_dark.png" alt="Dark Logo"/>
  </div>
  
  <span class="sidebar-brand-text">readfish 2024.2.0</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="toml.html">TOML files</a></li>
<li class="toctree-l1"><a class="reference internal" href="readfish.console.html">readfish console scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="guppy-params.html">Base-calling server parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="post-analysis.html">Readfish analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers-guide.html">Developer’s guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="readfish.html">API Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of API Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="readfish.html">API Reference</a></li>
<li class="toctree-l2"><a class="reference internal" href="readfish.plugins.html">readfish plugins API</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/getting-started.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="getting-started">
<h1>Getting Started<a class="headerlink" href="#getting-started" title="Link to this heading">¶</a></h1>
<p>Readfish is an adaptive sampling implementation for <a class="reference external" href="https://nanoporetech.com">ONT</a> sequencers.
So to begin with you will need an ONT sequencer, currently we support the MinION Mk1B, GridION, and PromethION platforms.
As we use live base-calling and alignment the computer that is controlling the sequencer will also need to be capable of base-calling; ideally this would use a GPU.</p>
<p>We recommend for alignment that you pre-prepare a minimap2 index (<code class="docutils literal notranslate"><span class="pre">.mmi</span></code>) file. This can be done by using the <code class="docutils literal notranslate"><span class="pre">minimap</span> <span class="pre">-d</span></code> flag. An example command -</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>minimap2<span class="w"> </span>-d<span class="w"> </span>&lt;output_index&gt;.mmi<span class="w"> </span>&lt;input_reference&gt;.fasta
</pre></div>
</div>
<p>Where <code class="docutils literal notranslate"><span class="pre">output_index</span></code> is the name of the index file to be output and <code class="docutils literal notranslate"><span class="pre">input_reference</span></code> is the name of the FASTA file that you want indexing.</p>
<p>This is then provided as the <code class="docutils literal notranslate"><span class="pre">fn_idx_in</span></code> value of the TOML file.</p>
<section id="testing-your-computer">
<h2>Testing your computer<a class="headerlink" href="#testing-your-computer" title="Link to this heading">¶</a></h2>
<!---
This is an aside, but we could have a 'readfish test-install' entry point.
This could be a complete series of test that starts with downloading a small bulk FAST5 file, setting up a simulation device and then starting/monitoring each test and eventually presenting the result.
-->
<p>To test readfish on your configuration we recommend first running a playback experiment to test unblock speed and then selection.</p>
<!-- begin-simulate -->
<!-- Adding a simulated position -->
<p>The following steps should all happen with a configuration (test) flow cell inserted into the target device.
A simulated device can also be created within MinKNOW, following these instructions. This assumes that you are runnning MinKNOW locally, using default ports. If this is not the case a developer API token is required on the commands as well, as well as setting the correct port.</p>
<p>If no test flow cell is available, a simulated device can be created within MinKNOW, following the below instructions.</p>
<!-- begin-simulate -->
<details style="margin-top: 10px; margin-bottom: 10px"><summary id="testing"><h3 style="display: inline">Adding a simulated position for testing</h3></summary>
<ol class="arabic">
<li><p>Linux</p>
<p>In the readfish virtual environment we created earlier:</p>
<ul class="simple">
<li><p>See help</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">python -m minknow_api.examples.manage_simulated_devices --help</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Add Minion position</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">python -m minknow_api.examples.manage_simulated_devices --add MS00000</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Add PromethION position</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">python -m minknow_api.examples.manage_simulated_devices --prom --add S0</span>
</pre></div>
</div>
</li>
<li><p>Mac</p>
<p>In the readfish virtual environment we created earlier:</p>
<ul class="simple">
<li><p>See help</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">python -m minknow_api.examples.manage_simulated_devices --help</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Add Minion position</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">python -m minknow_api.examples.manage_simulated_devices --add MS00000</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Add PromethION position</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">python -m minknow_api.examples.manage_simulated_devices --prom --add S0</span>
</pre></div>
</div>
</li>
</ol>
<p>As a back up it is possible to restart MinKNOW with a simulated device. This is done as follows:</p>
<ol class="arabic">
<li><p>Stop <code class="docutils literal notranslate"><span class="pre">minknow</span></code></p>
<p>On Linux:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cd /opt/ont/minknow/bin</span>
<span class="go">sudo systemctl stop minknow</span>
</pre></div>
</div>
</li>
<li><p>Start MinKNOW with a simulated device</p>
<p>On Linux</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">sudo ./mk_manager_svc -c /opt/ont/minknow/conf --simulated-minion-devices=1 &amp;</span>
</pre></div>
</div>
</li>
</ol>
<p>You <em>may</em> need to add the host <code class="docutils literal notranslate"><span class="pre">127.0.0.1</span></code> in the MinKNOW UI.</p>
</details>
<!-- end-simulate -->
<!-- #### Configuring bulk FAST5 file Playback -->
<details style="margin-top: 10px"><summary id="configuring-bulk-fast5-file"><h3 style="display: inline;">Configuring bulk FAST5 file Playback</h3></summary>
<p>Download an open access bulk FAST5 file, either <a class="reference external" href="https://s3.amazonaws.com/nanopore-human-wgs/bulkfile/PLSP57501_20170308_FNFAF14035_MN16458_sequencing_run_NOTT_Hum_wh1rs2_60428.fast5">R9.4.1 4khz</a> or <a class="reference external" href="https://s3.amazonaws.com/nanopore-human-wgs/bulkfile/GXB02001_20230509_1250_FAW79338_X3_sequencing_run_NA12878_B1_19382aa5_ef4362cd.fast5">R10 (5khz)</a>.
This file is 21Gb so make sure you have sufficient space.
A promethION bulkfile is also available but please note this is <a class="reference external" href="https://s3.amazonaws.com/nanopore-human-wgs/bulkfile/PC24B243_20220512_1516_PAK21362_3H_sequencing_run_NA12878_sheared20kb_3d5147fc.fast5">R10.4 4khz</a> and so will give slightly unexpected results on MinKNOW which assumes 5khz.
This file is approx 35Gb in size.</p>
<!-- begin-new-playback -->
<p>Previously to set up Playback using a pre-recorded bulk FAST5 file, it was necessary to edit the sequencing configuration file that MinKNOW uses. This is currently no longer the case. The “old method” steps are left after this section for reference only or if the direct playback from a bulk file option is removed in future.</p>
<p>To start sequencing using playback, simply begin setting up the run in the MinKNOW UI as you would usually.
Under Run Options you can select Simulated Playback and browse to the downloaded Bulk Fast5 file.</p>
<p><img alt="Run Options Screenshot" src="https://github.com/LooseLab/readfish/blob/247185a1bdcbe1275c55a6b4b1e2c7273213af91/docs/_static/images/simulated_playback_run_options.png?raw=true" /></p>
<!-- Included so these work in the github pages docs as well  -->
<!-- end-new-playback -->
<blockquote>
<div><p>[!NOTE]<br />
Note - The below instructions, whilst they will still work, are no longer required. They are left here for reference only. As of Minknow 5.7, it is possible to select a bulk FAST5 file for playback in the MinKNOW UI.</p>
</div></blockquote>
<!-- begin-obsolete -->
<details style="margin-top: 10px"><summary id="configuring-sequencing-toml"><h3 style="display: inline;">Old method Configuring bulk FAST5 file Playback</h3></summary>
To setup a simulation the sequencing configuration file that MinKNOW uses must be edited.
Steps:
<ol class="arabic">
<li><p>Download an open access bulkfile - either <a class="reference external" href="https://s3.amazonaws.com/nanopore-human-wgs/bulkfile/PLSP57501_20170308_FNFAF14035_MN16458_sequencing_run_NOTT_Hum_wh1rs2_60428.fast5">R9.4.1</a> or <a class="reference external" href="https://s3.amazonaws.com/nanopore-human-wgs/bulkfile/GXB02001_20230509_1250_FAW79338_X3_sequencing_run_NA12878_B1_19382aa5_ef4362cd.fast5">R10 (5khz)</a>. These files are approximately 21Gb so make sure you have plenty of space. The files are from NA12878 sequencing data using either R9.4.1 or R10.4 pores. Data is not barcoded and the libraries were ligation preps from DNA extracted from cell lines.</p></li>
<li><p>A promethION bulkfile is also available but please note this is <a class="reference external" href="https://s3.amazonaws.com/nanopore-human-wgs/bulkfile/PC24B243_20220512_1516_PAK21362_3H_sequencing_run_NA12878_sheared20kb_3d5147fc.fast5">R10.4, 4khz</a>, and so will give slightly unexpected results on MinKNOW which assumes 5khz.</p></li>
<li><p>Copy a sequencing TOML file to the <code class="docutils literal notranslate"><span class="pre">user_scripts</span></code> folder:</p>
<p>On Mac if your MinKNOW output directory is the default:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">mkdir -p /Library/MinKNOW/data/user_scripts/simulations</span>
<span class="go">cp /Applications/MinKNOW.app/Contents/Resources/conf/package/sequencing/sequencing_MIN106_DNA.toml /Library/MinKNOW/data/user_scripts/simulations/sequencing_MIN106_DNA_sim.toml</span>
</pre></div>
</div>
<p>On Linux:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">sudo mkdir -p /opt/ont/minknow/conf/package/sequencing/simulations</span>
<span class="go">cp /opt/ont/minknow/conf/package/sequencing/sequencing_MIN106_DNA.toml /opt/ont/minknow/conf/package/sequencing/simulations/sequencing_MIN106_DNA_sim.toml</span>
</pre></div>
</div>
</li>
<li><p>Edit the copied file to add the following line under the line that reads “<code class="docutils literal notranslate"><span class="pre">[custom_settings]</span></code>”:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>simulation = &quot;/full/path/to/your_bulk.FAST5&quot;
</pre></div>
</div>
<p>Change the text between the quotes to point to your downloaded bulk FAST5 file.</p>
 <!-- end-obsolete -->
</li>
<li><p>Optional, If running GUPPY in GPU mode, set the parameter <code class="docutils literal notranslate"><span class="pre">break_reads_after_seconds</span> <span class="pre">=</span> <span class="pre">1.0</span></code>
to <code class="docutils literal notranslate"><span class="pre">break_reads_after_seconds</span> <span class="pre">=</span> <span class="pre">0.4</span></code>. This results in a smaller read chunk. For R10.4 this is not required but can be tried. For adaptive sampling on PromethION, this should be left at 1 second.</p></li>
<li><p>In the MinKNOW GUI, right click on a sequencing position and select <code class="docutils literal notranslate"><span class="pre">Reload</span> <span class="pre">Scripts</span></code>.
Your version of MinKNOW will now playback the bulkfile rather than live sequencing.</p></li>
<li><p>Start a sequencing run as you would normally, selecting the corresponding flow
cell type to the edited script (here FLO-MIN106) as the flow cell type.</p></li>
</ol>
</details>
<p>Whichever instructions you followed, the run should start and immediately begin a mux scan. Let it run for around
five minutes after which your read length histogram should look as below:
<img alt="Control Image Screenshot" src="https://github.com/LooseLab/readfish/raw/main/docs/_static/images/PlaybackControlRun30Minutes.png?raw=true" /></p>
</details>
<details style="margin-top: 10px">
<summary id="testing-unblock-response"><h3 style="display: inline;">Testing unblock response</h3></summary>
<p>Now we shall test unblocking by running <code class="docutils literal notranslate"><span class="pre">readfish</span> <span class="pre">unblock-all</span></code> which will simply eject
every single read on the flow cell.</p>
<ol class="arabic">
<li><p>To do this run:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">readfish unblock-all --device &lt;YOUR_DEVICE_ID&gt; --experiment-name &quot;Testing readfish Unblock All&quot;</span>
</pre></div>
</div>
</li>
<li><p>Leave the run for a further 5 minutes and observe the read length histogram.
If unblocks are happening correctly you will see something like the below:
<img alt="Unblock All Screenshot" src="https://github.com/LooseLab/readfish/raw/main/docs/_static/images/PlaybackUnblockAll30minutes.png?raw=true" />
A closeup of the unblock peak shows reads being unblocked quickly:
<img alt="Closeup Unblock Image" src="https://github.com/LooseLab/readfish/raw/main/docs/_static/images/PlaybackUnblockAllCloseUp.png?raw=true" /></p></li>
</ol>
<p>If you are happy with the unblock response, move on to testing base-calling.</p>
<p>Note: The plots here are generated from running readfish unblock-all on an Apple Silicon laptop. The unblock response may be faster on a GPU server.</p>
</details>
<details style="margin-top: 10px">
<summary id="testing-basecalling-and-mapping"><h3 style="display: inline;">Testing base-calling and mapping</h3></summary>
<p>To test selective sequencing you must have access to either a
<a class="reference external" href="https://community.nanoporetech.com/downloads/guppy/release_notes">guppy basecall server</a> (&gt;=6.0.0) or a <a class="reference external" href="https://community.nanoporetech.com/downloads/dorado/release_notes">dorado basecall server</a>.</p>
<p>and a readfish TOML configuration file.</p>
<p>NOTE: guppy and dorado are used here interchangeably as the basecall server. Dorado is gradually replacing guppy. All readfish code is compatible with Guppy &gt;=6.0.0 and dorado &gt;=0.4.0</p>
<ol class="arabic">
<li><p>First make a local copy of the example TOML file:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">curl -O https://raw.githubusercontent.com/LooseLab/readfish/master/docs/_static/example_tomls/human_chr_selection.toml</span>
</pre></div>
</div>
</li>
<li><p>If on PromethION, edit the <code class="docutils literal notranslate"><span class="pre">mapper_settings.mappy</span></code> section to read:</p>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[mapper_settings.mappy-rs]</span>
</pre></div>
</div>
</li>
<li><p>If on MinKNOW core&gt;=5.9.0 and Dorado server version &gt;=7.3.9, edit the <code class="docutils literal notranslate"><span class="pre">basecaller</span></code> section to read:</p>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[caller_settings.dorado]</span>
</pre></div>
</div>
</li>
<li><p>Modify the <code class="docutils literal notranslate"><span class="pre">fn_idx_in</span></code> field in the file to be the full path to a <a class="reference external" href="https://github.com/lh3/minimap2">minimap2</a> index of the human genome.</p></li>
<li><p>Modify the <code class="docutils literal notranslate"><span class="pre">targets</span></code> fields for each condition to reflect the naming convention used in your index. This is the sequence name only, up to but not including any whitespace.
e.g. <code class="docutils literal notranslate"><span class="pre">&gt;chr1</span> <span class="pre">human</span> <span class="pre">chromosome</span> <span class="pre">1</span></code> would become <code class="docutils literal notranslate"><span class="pre">chr1</span></code>. If these names do not match, then target matching will fail.</p></li>
</ol>
<p>We can now validate this TOML file to see if it will be loaded correctly.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">readfish validate human_chr_selection.toml</span>
</pre></div>
</div>
<p>Errors with the configuration will be written to the terminal along with a text description of the conditions for the experiment as below.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>2023-10-05 15:29:18,934 readfish /home/adoni5/mambaforge/envs/readfish_dev/bin/readfish validate human_chr_selection.toml
2023-10-05 15:29:18,934 readfish command=&#39;validate&#39;
2023-10-05 15:29:18,934 readfish log_file=None
2023-10-05 15:29:18,934 readfish log_format=&#39;%(asctime)s %(name)s %(message)s&#39;
2023-10-05 15:29:18,934 readfish log_level=&#39;info&#39;
2023-10-05 15:29:18,934 readfish no_check_plugins=False
2023-10-05 15:29:18,934 readfish no_describe=False
2023-10-05 15:29:18,934 readfish prom=False
2023-10-05 15:29:18,934 readfish toml=&#39;human_chr_selection.toml&#39;
2023-10-05 15:29:18,934 readfish.validate eJydVk1v2zgQvetXEMqlxdryxyZAGyAHt0WKAk1TNNlTkBVoiZKIUKQiUonTX79vSEmW2zRo1/BBIkdvZt68GfKIXXV1zdunU3Z9efGZZUYXsmSFVIIVpmWt4GruZC3YlluRcaWkLmdM6FZmFR7JKDpi7tGwrGpNbayphWWv8MLWS8Z1ztar16zAFnOVYFVXc81KoWHGpGacWaDAWStKaXQCrOtK2v6V8aZREnjOMLiGC661UJbxrDXWesTHylCsyjxmTCiVRAOE2PG6wRYeQ1ZdK/KQVKc1xf4oXYUQ2be3yXGyYttO3e0zd8I6GHm8jxSvzJjjbSkc3LeC2UZkspCAzGUrMqeeKB+KyJlaBZx+AXA1UBiLEYiTZYwXeOD2dLo6s4B3M6FzPLVgjswokj6RGQyrdj1bzlbL5eyvOCYMvxQTbZ8KqsCa0h1Dm3n3AugIOHhB0iBy61+tzAVxwvvEZgyUbw1ICQHYUA5hBWu4c6LVoBJ84eta7gjeG0/TRki+qklmHzwHnr8NXJrCW20FKoUdofLAo9g1ikuNMPBhbbCSC8elGmBzk3U1UuCOBDEHWuVcY08XC2WMFYpvkxJ17LaJNAvINS+krRYUTFK5WpH7d710RTsqwaNFN2E1tcJRrW1Sdk3zdOurMv39639szuJgEY8UW2Y6oFaIRI8tIqglPpKhX5r3bXPoPChE81pEfdOdsTjXPG1XS5JjKt4k6/R4udw2Nj25q76nBbeONLHJ81ZA/T2j5eioz3FONQOLBe+UY7y3JiUFUyhENhkIXLi6WYSMFif4JdFgjFCeNyH/54jjHnm7pnMeVupcPsi844rmBWQTGhD/y6/Xny6/bD4jJu7bFVKivAcdZTQG7jvpBFMkQRLcN1GbB7xDE9T3ubR8SzrKxbYrU2U8UUo+iDQ4K+7joDFZamT//rDCNUbItML0/nKFvcW0wn6BlO0f5q3tc8FI8i4H56RSEFCgp3TmY++sSNhVZTqVU9MI6BQRnm+urjd+AGh2cfEpKnQq810KvSOxRQVKFjw3Wp4sPvTat4t30khNcwRpZRY6L+yiKv9+k2qTcuXAAk/K74nFuHRhA8MgvUjqWlLJLhtiA/X5ujmfVo5oDGl4N39G/+S7hhfk5ktXb5GgF6YvziAPsSP9vwpM0qEwUPl6fNsbEMNGq6fXkU4HnDN2HPng/LFwUGMUzQrxZ1PhiIOMJyvtPBw0IVA/fUaaw2n0QRRSS+dtkFfc6a0y2V2Mady0JhMij30KsXWmgSIzgVaA0GI/3DgBK0w8G0UksyPWf3RKMxGD0IBl7zarOn1HhNNo5o2jwwrzVRS06donoge7Nb8DqjZeSDkUahHZjMlEDMg+E20eB4d9wKfsn/DglUuMDAaHgQ+BDVZ9SFbcd6RqxETZ0jds/AZneEni8jn0NXc/HuX+3An3huFmkdux8s7fH/obAz2tkujmpi/O7W1Ec5KEh/tDSidzHNVSp73DM7ZCHliQdVczPYqw3+5J5CNf4yHGcxHVfLfHOSYcvnseJzQ0nUsjEMpB0WO6FTgeQRryxbghBYUJsUUvpRMXPPNjabhInLGb+Mv7dLlav10vk1V8SwX58bbhbyN7JqNwY0qNnxeH1Yvp+433QeE6Uov0x0TrL0Jebj3lZOI9JFGNg0L+CvBlRC9eh3vZFHvWX60cUwKHhd8afA3RFwV5G9ppmMO/HT3pRDq/CqQAPuTxPPQL2NSq/lu6L/YeLJIIm9AtGez9JBGmLjqCvIxDYPJ7MQltxmaazhqChOdnA/8NyIGWKeJP4jvA/gXiz7IPWqD7mQ16ZnvIyF/n0oNenFDyv3yEG+IeMvoPUvBL7w==
2023-10-05 15:29:18,937 readfish.validate Loaded TOML config without error
2023-10-05 15:29:18,937 readfish.validate Initialising Caller
2023-10-05 15:29:18,945 readfish.validate Caller initialised
2023-10-05 15:29:18,945 readfish.validate Initialising Aligner
2023-10-05 15:29:18,947 readfish.validate Aligner initialised
2023-10-05 15:29:18,948 readfish.validate Configuration description:
Region hum_test (control=False).
Region applies to section of flow cell (# = applied, . = not applied):

    ################################
    ################################
    ################################
    ################################
    ################################
    ################################
    ################################
    ################################

2023-10-05 15:29:18,948 readfish.validate Using the mappy plugin. Using reference: /home/adoni5/Documents/Bioinformatics/refs/hg38_no_alts.fa.gz.split/hg38_chr_M.mmi.

Region hum_test has targets on 1 contig, with 1 found in the provided reference.
This region has 2 total targets (+ve and -ve strands), covering approximately 100.00% of the genome.
</pre></div>
</div>
<ol class="arabic">
<li><p>If your toml file validates then run the following command:</p></li>
<li><div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">readfish targets --toml &lt;PATH_TO_TOML&gt; --device &lt;YOUR_DEVICE_ID&gt; --log-file test.log --experiment-name human_select_test</span>
</pre></div>
</div>
</li>
<li><p>In the terminal window you should see messages reporting the speed of mapping of the form:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>2023-10-05 15:24:03,910 readfish.targets MinKNOW is reporting PHASE_MUX_SCAN, waiting for PHASE_SEQUENCING to begin.
2023-10-05 15:25:48,150 readfish._read_until_client Protocol phase changed to PHASE_SEQUENCING
2023-10-05 15:25:48,724 readfish.targets 0494R/0.5713s; Avg: 0494R/0.5713s; Seq:0; Unb:494; Pro:0; Slow batches (&gt;1.00s): 0/1
2023-10-05 15:25:52,132 readfish.targets 0004R/0.1831s; Avg: 0249R/0.3772s; Seq:0; Unb:498; Pro:0; Slow batches (&gt;1.00s): 0/2
2023-10-05 15:25:52,600 readfish.targets 0122R/0.2494s; Avg: 0206R/0.3346s; Seq:0; Unb:620; Pro:0; Slow batches (&gt;1.00s): 0/3
2023-10-05 15:25:52,967 readfish.targets 0072R/0.2144s; Avg: 0173R/0.3046s; Seq:0; Unb:692; Pro:0; Slow batches (&gt;1.00s): 0/4
2023-10-05 15:25:53,349 readfish.targets 0043R/0.1932s; Avg: 0147R/0.2823s; Seq:0; Unb:735; Pro:0; Slow batches (&gt;1.00s): 0/5
2023-10-05 15:25:53,759 readfish.targets 0048R/0.2011s; Avg: 0130R/0.2688s; Seq:0; Unb:783; Pro:0; Slow batches (&gt;1.00s): 0/6
2023-10-05 15:25:54,206 readfish.targets 0126R/0.2458s; Avg: 0129R/0.2655s; Seq:0; Unb:909; Pro:0; Slow batches (&gt;1.00s): 0/7
2023-10-05 15:25:54,580 readfish.targets 0082R/0.2180s; Avg: 0123R/0.2595s; Seq:0; Unb:991; Pro:0; Slow batches (&gt;1.00s): 0/8
2023-10-05 15:25:54,975 readfish.targets 0053R/0.2110s; Avg: 0116R/0.2542s; Seq:0; Unb:1,044; Pro:0; Slow batches (&gt;1.00s): 0/9
2023-10-05 15:25:55,372 readfish.targets 0057R/0.2051s; Avg: 0110R/0.2492s; Seq:0; Unb:1,101; Pro:0; Slow batches (&gt;1.00s): 0/10
2023-10-05 15:25:55,817 readfish.targets 0135R/0.2467s; Avg: 0112R/0.2490s; Seq:0; Unb:1,236; Pro:0; Slow batches (&gt;1.00s): 0/11
2023-10-05 15:25:56,192 readfish.targets 0086R/0.2206s; Avg: 0110R/0.2466s; Seq:0; Unb:1,322; Pro:0; Slow batches (&gt;1.00s): 0/12
2023-10-05 15:25:56,588 readfish.targets 0060R/0.2138s; Avg: 0106R/0.2441s; Seq:0; Unb:1,382; Pro:0; Slow batches (&gt;1.00s): 0/13
2023-10-05 15:25:56,989 readfish.targets 0060R/0.2123s; Avg: 0103R/0.2418s; Seq:0; Unb:1,442; Pro:0; Slow batches (&gt;1.00s): 0/14
2023-10-05 15:25:57,429 readfish.targets 0133R/0.2502s; Avg: 0105R/0.2424s; Seq:0; Unb:1,575; Pro:0; Slow batches (&gt;1.00s): 0/15
2023-10-05 15:25:57,809 readfish.targets 0089R/0.2280s; Avg: 0104R/0.2415s; Seq:0; Unb:1,664; Pro:0; Slow batches (&gt;1.00s): 0/16
2023-10-05 15:25:58,210 readfish.targets 0059R/0.2247s; Avg: 0101R/0.2405s; Seq:0; Unb:1,723; Pro:0; Slow batches (&gt;1.00s): 0/17
^C2023-10-05 15:25:58,238 readfish.targets Keyboard interrupt received, stopping readfish
</pre></div>
</div>
</li>
</ol>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head text-left"><p>WARNING</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p><strong>Note: if these times are longer than the number of seconds specified in the break read chunk in the sequencing TOML, you will have performance issues. Contact us via github issues for support.</strong></p></td>
</tr>
</tbody>
</table>
</div>
<!-- start-heartbeat-log -->
<p>This log is a little dense at first. Moving from left to right, we have:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[Date Time] [Logger Name] [Batch Stats]; [Average Batch Stats]; [Count commands sent]; [Slow Batch Info]
</pre></div>
</div>
<p>Using the provided log as an example:</p>
<p>On 2023-10-05 at 15:25:56,989, the Readfish targets command logged a batch of read signal:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>- It saw 60 reads in the current batch.
- The batch took 0.2123 seconds.
- On average, batches are 103 reads, which are processed in 0.2418 seconds.
- Since the start, 0 reads were sequenced, 1,442 reads were unblocked, and 0 reads were asked to proceed.
- Out of 14 total batches processed, 0 were considered slow (took more than 1 second).
</pre></div>
</div>
<p>The important thing to note here is that the average batch time is less than the break read chunk time in the sequencing TOML.
The slow batch section will show the number of batches that were slower than break reads.
If the average is lower, or the slow batch count is high,  you will have performance issues.
Contact us via github issues for support.</p>
<!-- end-heartbeat-log -->
<p>If you are happy with the speed of mapping, move on to testing a selection.</p>
</details>
<details style="margin-top: 10px">
<summary id="testing-expected-results-from-a-selection-experiment"><h3 style="display: inline;">Testing expected results from a selection experiment.</h3></summary>
<p>The only way to test readfish on a playback run is to look at changes in read length for rejected vs accepted reads. To do this:</p>
<ol class="arabic">
<li><p>Start a fresh simulation run using the bulkfile provided above.</p></li>
<li><p>Restart the readfish command (as above):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">readfish targets --toml &lt;PATH_TO_TOML&gt; --device &lt;YOUR_DEVICE_ID&gt; --log-file test.log --experiment-name human_select_test</span>
</pre></div>
</div>
</li>
<li><p>Allow the run to proceed for at least 15 minutes (making sure you are writing out read data!).</p></li>
<li><p>After 15 minutes it should look something like this:
<img alt="Playback Unblock Image" src="https://github.com/LooseLab/readfish/raw/main/docs/_static/images/PlaybackRunTargeted.png?raw=true" />
If one zooms in on the unblock peak:
<img alt="Closeup Playback Unblock Image" src="https://github.com/LooseLab/readfish/raw/main/docs/_static/images/PlaybackRunTargetedUnblockPeak.png?raw=true" />
And if one zooms to exclude the unblock peak:
<img alt="Closeup Playback On Target Image" src="https://github.com/LooseLab/readfish/raw/main/docs/_static/images/PlaybackRunTargetedPeak.png?raw=true" />
NOTE: These simulations are also run on Apple Silicon - GPU platform performance may vary - please contact us via github issues for support.</p></li>
</ol>
 </details>
 <!-- /Testing expected results from a selection experiment. -->
 </details>
<details style="margin-top: 10px">
<summary id="Analysing-results-with-readfish-stats"><h3 style="display: inline;">Analysing results with readfish stats</h3></summary>
Once a run is complete, it can be analysed with the readfish stats command.
<p>HTML file output is optional.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">readfish stats --toml &lt;path/to/toml/file.toml&gt; --fastq-directory  &lt;path/to/run/folder&gt; --html &lt;filename&gt;</span>
</pre></div>
</div>
<p>Readfish stats will use the initial experiment configuration to analyse the final sequence data and output a formatted table to the screen.
The table is broken into two sections. For clarity these are shown individually below.</p>
<p>In the first table, the data is summarised by condition as defined in the TOML file.
In this example we have a single Region - “hum_test”.
The total number of reads is shown, along with the number of alignments broken down into On-Target and Off-Target.
In addition, we show yield, median read length and a summary of the number of targets.</p>
<table><tr><td style="font-weight: bold;color: #000000;text-align: left;">Condition</td><td style="font-weight: bold;color: #000000;text-align: left;">Reads</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Alignments</td><td colspan="4" style="font-weight: bold;color: #000000;text-align: left;">Yield</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Median read lengths</td><td style="font-weight: bold;color: #000000;text-align: left;">Number of targets</td><td style="font-weight: bold;color: #000000;text-align: left;">Percent target</td><td style="font-weight: bold;color: #000000;text-align: left;">Estimated coverage</td></tr><tr><td style="color: #555555;font-weight: bold;text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="text-align: left;">Total</td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="text-align: left;">Total</td><td style="text-align: left;">Ratio</td><td style="text-align: left;">On-target</td><td style="text-align: left;">Off-target</td><td style="text-align: left;">Combined</td><td style="text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;"></td></tr><tr><td style="color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">112,058</td><td style="color: #00aa00;text-align: left;">819 (0.73%)</td><td style="color: #aa0000;text-align: left;">111,239 (99.27%)</td><td style="color: #000000;text-align: left;">112,058</td><td style="color: #00aa00;text-align: left;">9.27 Mb (5.49%)</td><td style="color: #aa0000;text-align: left;">159.43 Mb (94.51%)</td><td style="color: #000000;text-align: left;">168.69 Mb</td><td style="color: #000000;text-align: left;">1:17.20</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">896 b</td><td style="color: #000000;text-align: left;">896 b</td><td style="color: #000000;text-align: left;">2</td><td style="color: #000000;text-align: left;">3.60%</td><td style="color: #000000;text-align: left;">0.08 X</td></tr><tr><td style="color: #555555;font-weight: bold;text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="text-align: left;">Total</td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="text-align: left;">Total</td><td style="text-align: left;">Ratio</td><td style="text-align: left;">On-target</td><td style="text-align: left;">Off-target</td><td style="text-align: left;">Combined</td><td style="text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;"></td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">Condition</td><td style="font-weight: bold;color: #000000;text-align: left;">Reads</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Alignments</td><td colspan="4" style="font-weight: bold;color: #000000;text-align: left;">Yield</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Median read lengths</td><td style="font-weight: bold;color: #000000;text-align: left;">Number of targets</td><td style="font-weight: bold;color: #000000;text-align: left;">Percent target</td><td style="font-weight: bold;color: #000000;text-align: left;">Estimated coverage</td></tr></table>
<p>The lower portion of the table shows the data broken down by contig in the reference (and so can be very long if using a complex reference!).
Again data are broken down by On and Off target. Read counts, yield, median and N50 read lengths are presented.
Finally we estimate the proportion of reads on target and an estimate of coverage.</p>
<p>In this experiment, we were targeting chromosomes 20 and 21. As this is a playback run there is no effect on yield but you can see a clear effect on read length.
The read length N50 and Median is higher for chromosomes 20 and 21 as expected. If running on more performant systems, the anticipated difference would be higher.</p>
<table><tr><td style="font-weight: bold;color: #000000;text-align: left;">Condition Name</td><td colspan="21" style="color: #000000;text-align: left;">hum_test</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">Condition</td><td style="font-weight: bold;color: #000000;text-align: left;">Contig</td><td style="font-weight: bold;color: #000000;text-align: left;">Contig Length</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Reads</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Alignments</td><td colspan="4" style="font-weight: bold;color: #000000;text-align: left;">Yield</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">Median read lengths</td><td colspan="3" style="font-weight: bold;color: #000000;text-align: left;">N50</td><td style="font-weight: bold;color: #000000;text-align: left;">Number of targets</td><td style="font-weight: bold;color: #000000;text-align: left;">Percent target</td><td style="font-weight: bold;color: #000000;text-align: left;">Estimated coverage</td></tr><tr><td style="color: #000000;font-weight: bold;text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;">Mapped</td><td style="text-align: left;">Unmapped</td><td style="font-weight: bold;text-align: left;">Total</td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="font-weight: bold;text-align: left;">Total</td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="font-weight: bold;text-align: left;">Total</td><td style="font-weight: bold;text-align: left;">Ratio</td><td style="text-align: left;">On-target</td><td style="text-align: left;">Off-target</td><td style="font-weight: bold;text-align: left;">Combined</td><td style="text-align: left;">On-Target</td><td style="text-align: left;">Off-Target</td><td style="font-weight: bold;text-align: left;">Total</td><td style="text-align: left;"></td><td style="text-align: left;"></td><td style="text-align: left;"></td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr1</td><td style="color: #000000;text-align: left;">248,956,422</td><td style="color: #00aa00;text-align: left;">10,015</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">10,015</td><td style="color: #00aa00;text-align: left;">6 (0.06%)</td><td style="color: #aa0000;text-align: left;">10,009 (99.94%)</td><td style="color: #000000;text-align: left;">10,015</td><td style="color: #00aa00;text-align: left;">48.65 Kb (0.37%)</td><td style="color: #aa0000;text-align: left;">13.03 Mb (99.63%)</td><td style="color: #000000;text-align: left;">13.08 Mb</td><td style="color: #000000;text-align: left;">1:267.87</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">891 b</td><td style="color: #000000;text-align: left;">891 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.35 Kb</td><td style="color: #000000;text-align: left;">1.35 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr2</td><td style="color: #000000;text-align: left;">242,193,529</td><td style="color: #00aa00;text-align: left;">8,825</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">8,825</td><td style="color: #00aa00;text-align: left;">9 (0.10%)</td><td style="color: #aa0000;text-align: left;">8,816 (99.90%)</td><td style="color: #000000;text-align: left;">8,825</td><td style="color: #00aa00;text-align: left;">47.36 Kb (0.36%)</td><td style="color: #aa0000;text-align: left;">13.05 Mb (99.64%)</td><td style="color: #000000;text-align: left;">13.09 Mb</td><td style="color: #000000;text-align: left;">1:275.51</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">894 b</td><td style="color: #000000;text-align: left;">894 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.49 Kb</td><td style="color: #000000;text-align: left;">1.49 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr3</td><td style="color: #000000;text-align: left;">198,295,559</td><td style="color: #00aa00;text-align: left;">8,005</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">8,005</td><td style="color: #00aa00;text-align: left;">6 (0.07%)</td><td style="color: #aa0000;text-align: left;">7,999 (99.93%)</td><td style="color: #000000;text-align: left;">8,005</td><td style="color: #00aa00;text-align: left;">193.03 Kb (1.73%)</td><td style="color: #aa0000;text-align: left;">10.98 Mb (98.27%)</td><td style="color: #000000;text-align: left;">11.17 Mb</td><td style="color: #000000;text-align: left;">1:56.86</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">893 b</td><td style="color: #000000;text-align: left;">893 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.42 Kb</td><td style="color: #000000;text-align: left;">1.42 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr4</td><td style="color: #000000;text-align: left;">190,214,555</td><td style="color: #00aa00;text-align: left;">7,381</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">7,381</td><td style="color: #00aa00;text-align: left;">30 (0.41%)</td><td style="color: #aa0000;text-align: left;">7,351 (99.59%)</td><td style="color: #000000;text-align: left;">7,381</td><td style="color: #00aa00;text-align: left;">861.07 Kb (7.29%)</td><td style="color: #aa0000;text-align: left;">10.95 Mb (92.71%)</td><td style="color: #000000;text-align: left;">11.81 Mb</td><td style="color: #000000;text-align: left;">1:12.72</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">917 b</td><td style="color: #000000;text-align: left;">917 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.60 Kb</td><td style="color: #000000;text-align: left;">1.60 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr5</td><td style="color: #000000;text-align: left;">181,538,259</td><td style="color: #00aa00;text-align: left;">7,545</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">7,545</td><td style="color: #00aa00;text-align: left;">5 (0.07%)</td><td style="color: #aa0000;text-align: left;">7,540 (99.93%)</td><td style="color: #000000;text-align: left;">7,545</td><td style="color: #00aa00;text-align: left;">50.70 Kb (0.50%)</td><td style="color: #aa0000;text-align: left;">10.18 Mb (99.50%)</td><td style="color: #000000;text-align: left;">10.23 Mb</td><td style="color: #000000;text-align: left;">1:200.68</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">896 b</td><td style="color: #000000;text-align: left;">896 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.40 Kb</td><td style="color: #000000;text-align: left;">1.40 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr6</td><td style="color: #000000;text-align: left;">170,805,979</td><td style="color: #00aa00;text-align: left;">5,808</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">5,808</td><td style="color: #00aa00;text-align: left;">9 (0.15%)</td><td style="color: #aa0000;text-align: left;">5,799 (99.85%)</td><td style="color: #000000;text-align: left;">5,808</td><td style="color: #00aa00;text-align: left;">116.44 Kb (1.35%)</td><td style="color: #aa0000;text-align: left;">8.53 Mb (98.65%)</td><td style="color: #000000;text-align: left;">8.65 Mb</td><td style="color: #000000;text-align: left;">1:73.28</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">905 b</td><td style="color: #000000;text-align: left;">905 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.49 Kb</td><td style="color: #000000;text-align: left;">1.49 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr7</td><td style="color: #000000;text-align: left;">159,345,973</td><td style="color: #00aa00;text-align: left;">6,383</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">6,383</td><td style="color: #00aa00;text-align: left;">2 (0.03%)</td><td style="color: #aa0000;text-align: left;">6,381 (99.97%)</td><td style="color: #000000;text-align: left;">6,383</td><td style="color: #00aa00;text-align: left;">26.06 Kb (0.29%)</td><td style="color: #aa0000;text-align: left;">9.11 Mb (99.71%)</td><td style="color: #000000;text-align: left;">9.14 Mb</td><td style="color: #000000;text-align: left;">1:349.59</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">895 b</td><td style="color: #000000;text-align: left;">895 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.44 Kb</td><td style="color: #000000;text-align: left;">1.44 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr8</td><td style="color: #000000;text-align: left;">145,138,636</td><td style="color: #00aa00;text-align: left;">5,208</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">5,208</td><td style="color: #00aa00;text-align: left;">1 (0.02%)</td><td style="color: #aa0000;text-align: left;">5,207 (99.98%)</td><td style="color: #000000;text-align: left;">5,208</td><td style="color: #00aa00;text-align: left;">285 b (0.00%)</td><td style="color: #aa0000;text-align: left;">7.43 Mb (100.00%)</td><td style="color: #000000;text-align: left;">7.43 Mb</td><td style="color: #000000;text-align: left;">1:26061.60</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">892 b</td><td style="color: #000000;text-align: left;">892 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.44 Kb</td><td style="color: #000000;text-align: left;">1.44 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr9</td><td style="color: #000000;text-align: left;">138,394,717</td><td style="color: #00aa00;text-align: left;">4,253</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">4,253</td><td style="color: #00aa00;text-align: left;">23 (0.54%)</td><td style="color: #aa0000;text-align: left;">4,230 (99.46%)</td><td style="color: #000000;text-align: left;">4,253</td><td style="color: #00aa00;text-align: left;">91.15 Kb (1.50%)</td><td style="color: #aa0000;text-align: left;">6.00 Mb (98.50%)</td><td style="color: #000000;text-align: left;">6.09 Mb</td><td style="color: #000000;text-align: left;">1:65.85</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">899 b</td><td style="color: #000000;text-align: left;">899 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.46 Kb</td><td style="color: #000000;text-align: left;">1.46 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr10</td><td style="color: #000000;text-align: left;">133,797,422</td><td style="color: #00aa00;text-align: left;">4,424</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">4,424</td><td style="color: #00aa00;text-align: left;">15 (0.34%)</td><td style="color: #aa0000;text-align: left;">4,409 (99.66%)</td><td style="color: #000000;text-align: left;">4,424</td><td style="color: #00aa00;text-align: left;">95.02 Kb (1.37%)</td><td style="color: #aa0000;text-align: left;">6.86 Mb (98.63%)</td><td style="color: #000000;text-align: left;">6.95 Mb</td><td style="color: #000000;text-align: left;">1:72.16</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">915 b</td><td style="color: #000000;text-align: left;">915 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.56 Kb</td><td style="color: #000000;text-align: left;">1.56 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr11</td><td style="color: #000000;text-align: left;">135,086,622</td><td style="color: #00aa00;text-align: left;">5,349</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">5,349</td><td style="color: #00aa00;text-align: left;">1 (0.02%)</td><td style="color: #aa0000;text-align: left;">5,348 (99.98%)</td><td style="color: #000000;text-align: left;">5,349</td><td style="color: #00aa00;text-align: left;">287 b (0.00%)</td><td style="color: #aa0000;text-align: left;">6.89 Mb (100.00%)</td><td style="color: #000000;text-align: left;">6.89 Mb</td><td style="color: #000000;text-align: left;">1:23997.50</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">896 b</td><td style="color: #000000;text-align: left;">896 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.35 Kb</td><td style="color: #000000;text-align: left;">1.35 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr12</td><td style="color: #000000;text-align: left;">133,275,309</td><td style="color: #00aa00;text-align: left;">5,508</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">5,508</td><td style="color: #00aa00;text-align: left;">3 (0.05%)</td><td style="color: #aa0000;text-align: left;">5,505 (99.95%)</td><td style="color: #000000;text-align: left;">5,508</td><td style="color: #00aa00;text-align: left;">2.63 Kb (0.03%)</td><td style="color: #aa0000;text-align: left;">7.59 Mb (99.97%)</td><td style="color: #000000;text-align: left;">7.59 Mb</td><td style="color: #000000;text-align: left;">1:2888.96</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">893 b</td><td style="color: #000000;text-align: left;">893 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.40 Kb</td><td style="color: #000000;text-align: left;">1.40 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr13</td><td style="color: #000000;text-align: left;">114,364,328</td><td style="color: #00aa00;text-align: left;">3,414</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">3,414</td><td style="color: #00aa00;text-align: left;">8 (0.23%)</td><td style="color: #aa0000;text-align: left;">3,406 (99.77%)</td><td style="color: #000000;text-align: left;">3,414</td><td style="color: #00aa00;text-align: left;">85.71 Kb (1.80%)</td><td style="color: #aa0000;text-align: left;">4.69 Mb (98.20%)</td><td style="color: #000000;text-align: left;">4.77 Mb</td><td style="color: #000000;text-align: left;">1:54.67</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">900 b</td><td style="color: #000000;text-align: left;">900 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.43 Kb</td><td style="color: #000000;text-align: left;">1.43 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr14</td><td style="color: #000000;text-align: left;">107,043,718</td><td style="color: #00aa00;text-align: left;">3,541</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">3,541</td><td style="color: #00aa00;text-align: left;">12 (0.34%)</td><td style="color: #aa0000;text-align: left;">3,529 (99.66%)</td><td style="color: #000000;text-align: left;">3,541</td><td style="color: #00aa00;text-align: left;">244.18 Kb (4.79%)</td><td style="color: #aa0000;text-align: left;">4.86 Mb (95.21%)</td><td style="color: #000000;text-align: left;">5.10 Mb</td><td style="color: #000000;text-align: left;">1:19.90</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">892 b</td><td style="color: #000000;text-align: left;">892 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.42 Kb</td><td style="color: #000000;text-align: left;">1.42 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr15</td><td style="color: #000000;text-align: left;">101,991,189</td><td style="color: #00aa00;text-align: left;">3,033</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">3,033</td><td style="color: #00aa00;text-align: left;">3 (0.10%)</td><td style="color: #aa0000;text-align: left;">3,030 (99.90%)</td><td style="color: #000000;text-align: left;">3,033</td><td style="color: #00aa00;text-align: left;">4.29 Kb (0.11%)</td><td style="color: #aa0000;text-align: left;">3.79 Mb (99.89%)</td><td style="color: #000000;text-align: left;">3.80 Mb</td><td style="color: #000000;text-align: left;">1:883.07</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">867 b</td><td style="color: #000000;text-align: left;">867 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.31 Kb</td><td style="color: #000000;text-align: left;">1.31 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr16</td><td style="color: #000000;text-align: left;">90,338,345</td><td style="color: #00aa00;text-align: left;">3,276</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">3,276</td><td style="color: #00aa00;text-align: left;">1 (0.03%)</td><td style="color: #aa0000;text-align: left;">3,275 (99.97%)</td><td style="color: #000000;text-align: left;">3,276</td><td style="color: #00aa00;text-align: left;">1.97 Kb (0.04%)</td><td style="color: #aa0000;text-align: left;">4.51 Mb (99.96%)</td><td style="color: #000000;text-align: left;">4.51 Mb</td><td style="color: #000000;text-align: left;">1:2294.28</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">900 b</td><td style="color: #000000;text-align: left;">900 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.41 Kb</td><td style="color: #000000;text-align: left;">1.41 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr17</td><td style="color: #000000;text-align: left;">83,257,441</td><td style="color: #00aa00;text-align: left;">3,378</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">3,378</td><td style="color: #00aa00;text-align: left;">10 (0.30%)</td><td style="color: #aa0000;text-align: left;">3,368 (99.70%)</td><td style="color: #000000;text-align: left;">3,378</td><td style="color: #00aa00;text-align: left;">16.81 Kb (0.36%)</td><td style="color: #aa0000;text-align: left;">4.72 Mb (99.64%)</td><td style="color: #000000;text-align: left;">4.73 Mb</td><td style="color: #000000;text-align: left;">1:280.52</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">907 b</td><td style="color: #000000;text-align: left;">907 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.43 Kb</td><td style="color: #000000;text-align: left;">1.43 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr18</td><td style="color: #000000;text-align: left;">80,373,285</td><td style="color: #00aa00;text-align: left;">3,158</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">3,158</td><td style="color: #00aa00;text-align: left;">3 (0.09%)</td><td style="color: #aa0000;text-align: left;">3,155 (99.91%)</td><td style="color: #000000;text-align: left;">3,158</td><td style="color: #00aa00;text-align: left;">186.59 Kb (4.06%)</td><td style="color: #aa0000;text-align: left;">4.41 Mb (95.94%)</td><td style="color: #000000;text-align: left;">4.59 Mb</td><td style="color: #000000;text-align: left;">1:23.61</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">899 b</td><td style="color: #000000;text-align: left;">899 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.47 Kb</td><td style="color: #000000;text-align: left;">1.47 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr19</td><td style="color: #000000;text-align: left;">58,617,616</td><td style="color: #00aa00;text-align: left;">2,110</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">2,110</td><td style="color: #00aa00;text-align: left;">0 (0.00%)</td><td style="color: #aa0000;text-align: left;">2,110 (100.00%)</td><td style="color: #000000;text-align: left;">2,110</td><td style="color: #00aa00;text-align: left;">0 b (0.00%)</td><td style="color: #aa0000;text-align: left;">2.53 Mb (100.00%)</td><td style="color: #000000;text-align: left;">2.53 Mb</td><td style="color: #000000;text-align: left;">0:0.00</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">857 b</td><td style="color: #000000;text-align: left;">857 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.27 Kb</td><td style="color: #000000;text-align: left;">1.27 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr20</td><td style="color: #000000;text-align: left;">64,444,167</td><td style="color: #00aa00;text-align: left;">370</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">370</td><td style="color: #00aa00;text-align: left;">370 (100.00%)</td><td style="color: #aa0000;text-align: left;">0 (0.00%)</td><td style="color: #000000;text-align: left;">370</td><td style="color: #00aa00;text-align: left;">3.60 Mb (100.00%)</td><td style="color: #aa0000;text-align: left;">0 b (0.00%)</td><td style="color: #000000;text-align: left;">3.60 Mb</td><td style="color: #000000;text-align: left;">1:0.00</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">2.88 Kb</td><td style="color: #000000;text-align: left;">2.88 Kb</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">32.28 Kb</td><td style="color: #000000;text-align: left;">32.28 Kb</td><td style="color: #000000;text-align: left;">1</td><td style="color: #000000;text-align: left;">100.00%</td><td style="color: #000000;text-align: left;">0.06 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr21</td><td style="color: #000000;text-align: left;">46,709,983</td><td style="color: #00aa00;text-align: left;">265</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">265</td><td style="color: #00aa00;text-align: left;">265 (100.00%)</td><td style="color: #aa0000;text-align: left;">0 (0.00%)</td><td style="color: #000000;text-align: left;">265</td><td style="color: #00aa00;text-align: left;">3.06 Mb (100.00%)</td><td style="color: #aa0000;text-align: left;">0 b (0.00%)</td><td style="color: #000000;text-align: left;">3.06 Mb</td><td style="color: #000000;text-align: left;">1:0.00</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">2.63 Kb</td><td style="color: #000000;text-align: left;">2.63 Kb</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">33.54 Kb</td><td style="color: #000000;text-align: left;">33.54 Kb</td><td style="color: #000000;text-align: left;">1</td><td style="color: #000000;text-align: left;">100.00%</td><td style="color: #000000;text-align: left;">0.07 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chr22</td><td style="color: #000000;text-align: left;">50,818,468</td><td style="color: #00aa00;text-align: left;">1,741</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">1,741</td><td style="color: #00aa00;text-align: left;">28 (1.61%)</td><td style="color: #aa0000;text-align: left;">1,713 (98.39%)</td><td style="color: #000000;text-align: left;">1,741</td><td style="color: #00aa00;text-align: left;">421.99 Kb (14.61%)</td><td style="color: #aa0000;text-align: left;">2.47 Mb (85.39%)</td><td style="color: #000000;text-align: left;">2.89 Mb</td><td style="color: #000000;text-align: left;">1:5.85</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">922 b</td><td style="color: #000000;text-align: left;">922 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.63 Kb</td><td style="color: #000000;text-align: left;">1.63 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chrM</td><td style="color: #000000;text-align: left;">16,569</td><td style="color: #00aa00;text-align: left;">19</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">19</td><td style="color: #00aa00;text-align: left;">0 (0.00%)</td><td style="color: #aa0000;text-align: left;">19 (100.00%)</td><td style="color: #000000;text-align: left;">19</td><td style="color: #00aa00;text-align: left;">0 b (0.00%)</td><td style="color: #aa0000;text-align: left;">16.82 Kb (100.00%)</td><td style="color: #000000;text-align: left;">16.82 Kb</td><td style="color: #000000;text-align: left;">0:0.00</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">774 b</td><td style="color: #000000;text-align: left;">774 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.11 Kb</td><td style="color: #000000;text-align: left;">1.11 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chrX</td><td style="color: #000000;text-align: left;">156,040,895</td><td style="color: #00aa00;text-align: left;">5,636</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">5,636</td><td style="color: #00aa00;text-align: left;">5 (0.09%)</td><td style="color: #aa0000;text-align: left;">5,631 (99.91%)</td><td style="color: #000000;text-align: left;">5,636</td><td style="color: #00aa00;text-align: left;">3.19 Kb (0.04%)</td><td style="color: #aa0000;text-align: left;">7.46 Mb (99.96%)</td><td style="color: #000000;text-align: left;">7.46 Mb</td><td style="color: #000000;text-align: left;">1:2336.71</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">905 b</td><td style="color: #000000;text-align: left;">905 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">1.38 Kb</td><td style="color: #000000;text-align: left;">1.38 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">chrY</td><td style="color: #000000;text-align: left;">57,227,415</td><td style="color: #00aa00;text-align: left;">116</td><td style="color: #aa0000;text-align: left;">0</td><td style="color: #000000;text-align: left;">116</td><td style="color: #00aa00;text-align: left;">4 (3.45%)</td><td style="color: #aa0000;text-align: left;">112 (96.55%)</td><td style="color: #000000;text-align: left;">116</td><td style="color: #00aa00;text-align: left;">117.28 Kb (27.65%)</td><td style="color: #aa0000;text-align: left;">306.90 Kb (72.35%)</td><td style="color: #000000;text-align: left;">424.19 Kb</td><td style="color: #000000;text-align: left;">1:2.62</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">989 b</td><td style="color: #000000;text-align: left;">989 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">28.59 Kb</td><td style="color: #000000;text-align: left;">28.59 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr><tr><td style="font-weight: bold;color: #000000;text-align: left;">hum_test</td><td style="color: #000000;text-align: left;">unmapped</td><td style="color: #000000;text-align: left;">0</td><td style="color: #00aa00;text-align: left;">0</td><td style="color: #aa0000;text-align: left;">3,297</td><td style="color: #000000;text-align: left;">3,297</td><td style="color: #00aa00;text-align: left;">0 (0.00%)</td><td style="color: #aa0000;text-align: left;">3,297 (100.00%)</td><td style="color: #000000;text-align: left;">3,297</td><td style="color: #00aa00;text-align: left;">0 b (0.00%)</td><td style="color: #aa0000;text-align: left;">9.10 Mb (100.00%)</td><td style="color: #000000;text-align: left;">9.10 Mb</td><td style="color: #000000;text-align: left;">0:0.00</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">508 b</td><td style="color: #000000;text-align: left;">508 b</td><td style="color: #00aa00;text-align: left;">0 b</td><td style="color: #aa0000;text-align: left;">16.81 Kb</td><td style="color: #000000;text-align: left;">16.81 Kb</td><td style="color: #000000;text-align: left;">0</td><td style="color: #000000;text-align: left;">0.00%</td><td style="color: #000000;text-align: left;">0.00 X</td></tr></table>
</details>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="toml.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">TOML files</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Loose Lab
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Getting Started</a><ul>
<li><a class="reference internal" href="#testing-your-computer">Testing your computer</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=0af83669"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=4e2eecee"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    </body>
</html>